### DISCUSSION TITLE
Test fail Received unexpected action without useful logs

### DISCUSSION THREAD

**Discussion by saroar** [pointfreeco/swift-composable-architecture#867] [General]
logs don't have any use full info so I can fixed my test case 

`
Received unexpected action: …

      EventFormAction.eventsResponse(
        Result.success(
          EventResponse.Item(
            _id: "5fbfe53675a93bda87c7cb16",
            name: "Cool :)",
            categories: "General",
            imageUrl: nil,
            duration: 14400,
            distance: nil,
            isActive: true,
            conversationsId: "5fbfe5361cdd72e23297914a",
            addressName: "8к1литД улица Вавиловых , Saint Petersburg",
            details: nil,
            type: "Point",
            sponsored: false,
            overlay: false,
            coordinates: […],
            regionRadius: 1000.0,
            createdAt: Date(1970-01-01T00:00:00.000Z),
            updatedAt: Date(1970-01-01T00:00:00.000Z)
          )
        )
      )

(Expected: −, Received: +)
`
https://github.com/AddaMeSPB/AddaMeIOS/blob/feature/git_workflow/AddameSPM/Tests/EventFormViewTests/EventFormViewTests.swift#L128
<img width="1437" alt="Screenshot 2021-10-21 at 12 21 32" src="https://user-images.githubusercontent.com/8770772/138249295-3c68c590-5a6a-4129-ae0c-433554a13c9e.png">

**Comment by stephencelis** [comment]
@saroar We should look into providing a better error here, but it looks like `EventResponse.Item` is an `NSObject` without a custom `==` check, so it's using object identity for the comparison. Can `EventResponse.Item` be a struct, instead?

In general we'd recommend sticking with value types wherever possible in TCA state and actions.

**Reply by iampatbrown** [reply]
@stephencelis is it worth doing something like this?
```swift
diff(expectedAction, receivedAction, format: .proportional)
  .map {
    """
    \($0.indent(by: 4))

    \($0.contains("\n\u{2212}") || $0.contains("\n+")
      ? "(Expected: −, Received: +)"
      : "Warning: No difference was detected. This can occur when using reference types.")
    """
  }
```
or would a more specific error be preferable (most likely requiring a custom diff)?

**Reply by saroar** [reply]
the reason was use class becz of this
<img width="1095" alt="Screenshot 2021-10-22 at 10 20 29" src="https://user-images.githubusercontent.com/8770772/138410894-3ab144a2-dd51-47f7-a096-4ce334d90e74.png">

or there is better way ?
https://github.com/AddaMeSPB/AddaMeIOS/blob/feature/git_workflow/AddameSPM/Sources/SharedModels/Event/Event.swift#L170

**Reply by stephencelis** [reply]
> or would a more specific error be preferable (most likely requiring a custom diff)?

I think the behavior is wrong in swift-custom-dump (it should always print a difference). I've taken one step in improving things here:

https://github.com/pointfreeco/swift-custom-dump/pull/29

**Reply by iampatbrown** [reply]
> I think the behavior is wrong in swift-custom-dump

Okay awesome. I initially started modifying swift-custom-dump, but then made the incorrect assumption that it was done that way deliberately.
